### ISSUE TITLE
filter(None, ...) with False input has false type

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#479]
**Describe the bug**
When filtering an iterable of `False` literals and other type values, the return type includes `bool`.  When combined with integers, the `int` type is - oddly - not in the output.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That the revealed type of `generator_of_ints` would be `Iterator[int]`.

**Screenshots or Code**
```py
from typing import List, Literal, Union

list_of_bools_and_ints: List[Union[Literal[False], int]] = []
generator_of_ints = filter(None, list_of_bools_and_ints)
reveal_type(generator_of_ints)
```

**VS Code extension or command-line**
extension, 1.1.18

**Comment by erictraut** [comment]
Thanks for the bug report. You're correct — the type should evaluate to Iterator[int]. This was a bug in Pyright's TypeVar matching logic. It will be fixed in the next release.

**Comment by erictraut** [comment]
This is now fixed in 1.1.19, which I just published.

**Comment by layday** [comment]
I'm still seeing this with other types. If I replace `int` with `str`:

```py
from typing import List, Literal, Union

list_of_bools_and_ints: List[Union[Literal[False], str]] = []
generator_of_ints = filter(None, list_of_bools_and_ints)
reveal_type(generator_of_ints)
```

... I get:

> Type of 'generator_of_ints' is 'Iterator[Union[str, bool]]'

But this should be `Iterator[str]`.

If I throw `None` into the mix:

```py
from typing import List, Literal, Union

list_of_bools_and_ints: List[Union[None, Literal[False], str]] = []
generator_of_ints = filter(None, list_of_bools_and_ints)
reveal_type(generator_of_ints)
```

... it becomes:

> Type of 'generator_of_ints' is 'Iterator[Optional[Union[str, bool]]]'

But again this should be `Iterator[str]`.  The same logic that applies to `if` expressions should also apply here.

**Comment by erictraut** [comment]
The `filter` case is not identical to a for loop. There are important differences.

I think that `Iterator[Union[str, bool]]` is the correct answer. The reason that you don't get `Iterator[Union[int, bool]]` in your previous example is that `bool` is a subclass of `int`, so the TypeVar matching logic collapses it int just `int`. But `str` and `bool` have no inheritance relationship, so the union must be retained during TypeVar matching.

There's an argument to be made that you should get `Iterator[Union[Literal[False], int]]`, but the TypeVar logic needs to strip literal types when it's solving for TypeVars, so `Literal[False]`, so Literal[False] gets translated to `bool`. That's why you end up with `Iterator[Union[str, bool]]`.

The for loop makes use of type constraint logic. The type constraint behavior is not formally specified in PEP 484, but Pyright's implementation does honor Literal[False] and Literal[True] when constraining (narrowing) types based on conditional statements.

**Comment by layday** [comment]
Syntactically it might not be identical, but the output is identical, is it not?  If I run my list of `False`s and a variety of strings through `filter(None)` or if I use a generator expression, the result will be the same and it won't include any booleans.

**Comment by layday** [comment]
Perhaps unrelated, but the latest release also introduced a regression where the return value is derived from the filter function, e.g. 

```py
def foo(value: object) -> bool:
    ...

reveal_type(filter(foo, ['b', 'a', 'r']))
```

produces:

> Type of 'filter(foo, ['b', 'a', 'r'])' is 'Iterator[object]'

**Comment by erictraut** [comment]
>Syntactically it might not be identical, but the output is identical, is it not?
It might be, but a type checker needs to base its decisions on the information it's given. In this case, it's using the type definition of the `filter` function provided in the stdlib type stub and applying the rules set forth in PEP 484.

And yes, you're correct that I introduced a regression in the recent release — or more specifically, uncovered a latent bug that was previously masked. I'll file a separate bug report on that issue. Thanks for pointing it out.

**Comment by layday** [comment]
Looking at the typeshed `filter` has an overloaded variant for `None` which (I think) is intended to weed out `None`s from the output:

```py
@overload
def filter(__function: None, __iterable: Iterable[Optional[_T]]) -> Iterator[_T]: ...
```

To me this looks ambiguous.  The optional translates to a union of `_T` and `None` and since `_T` is not constrained it can very well be `None`.  Perhaps there is something in one of the many typing PEPs that explains it.

**Comment by erictraut** [comment]
Yeah, I noticed that too. I agree that it's somewhat ambiguous. The way I interpret it is that `Optional[_T]` is equivalent to just `_T` because the TypeVar is unconstrained and can be a union of anything, including None. I don't recall any of the typing PEPs calling this out as a special case. It's possible that mypy treats it as a special case.

Regardless, that doesn't speak to your example above because `None` will never match `Literal[False]`. Both types evaluate to false when used in a conditional expression, but they aren't interchangeable.

**Comment by layday** [comment]
No, indeed, I was more thinking if the stub should be expanded to include all falsy literals now that `Literal` is a thing but if a union with an unconstrained `_T` is just `_T` in Pyright, then I don't suppose there's any point.  Unfortunate because it does kind of reduce the usefulness of `filter`.

**Comment by erictraut** [comment]
There are very few types that are statically provable to be either falsy or truthy. Many types evaluate to either falsy or truthy depending on their values. 

I'd argue that the case that you're trying to handle is an edge case, not the norm. The examples that you gave above (`Union[Literal[False], int]` and `Union[Literal[False], str]`) are inconsistent in that you are specifically calling out `Literal[False]` but not `Literal[0]` or `Literal[""]`, which are also falsy.

**Comment by layday** [comment]
Right, I see what you mean.  However, the type is accurately inferred in if expressions for known falsy literals, including False, None and the empty tuple.  I was wondering if this shouldn't also apply to `filter(None)`, owing to its ubiquitousness.  The following idiom is very common in Python:

```py
filtered_iterable = filter(None, map(some_func, an_iterable))
```

If the type system is not expressive enough to capture the relationship between `None` and the output, I'm wondering whether this should be special-cased to match the behaviour of the [equivalent](https://docs.python.org/3/library/functions.html#filter) expression: `(i for i in an_iterable if i)`.  Yeah, it's a kludge, but it might be a useful one.
